pkg/operator/certupdater/cert_updater.go

```diff
@@ -0,0 +1,225 @@
+
+			select {
+			case <-time.After(cu.pollingInterval):
+				// Noop
+			case <-ctx.Done():
+				done = true
+			}
+		}
+	}()
+	return nil
+}
+
+func (cu *CertUpdater) poll(ctx context.Context) error {
+	if cu.source == nil {
+		return fmt.Errorf("certificate source not set")
+	}
+
+	cert, ca, err := cu.source(ctx)
+	if err != nil {
+		return fmt.Errorf("cert source: %w", err)
+	}
+	cu.mu.Lock()
+	cu.currentCert = &cert
+	cu.currentCA = ca
+	cu.mu.Unlock()
+
+	return nil
+}
+
+// GetCA allows access to CA Bundle, if applicable
+func (cu *CertUpdater) GetCA() (*x509.Certificate, error) {
+	cu.mu.RLock()
+	defer cu.mu.RUnlock()
+	return cu.currentCA, nil
+}
+
+// GetCertificate implements tls.Config.GetCertificate. Certificates are updated asynchronously.
+func (cu *CertUpdater) GetCertificate(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	cu.mu.RLock()
+	defer cu.mu.RUnlock()
+	return cu.currentCert, nil
+}
+
+// CertSource defines the common signature of functions that return certificates
+type CertSource func(ctx context.Context) (webhook tls.Certificate, ca *x509.Certificate, err error)
+
+// SourceBase64 sources certificates from base64 strings
+func SourceBase64(certString, keyString string, optionalCAString string) (CertSource, error) {
+	cert, err := base64.StdEncoding.DecodeString(certString)
+	if err != nil {
+		return nil, err
+	}
+	key, err := base64.StdEncoding.DecodeString(keyString)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use cert as CA, if not provided
+	if optionalCAString == "" {
+		return sourcePEM(cert, key, cert)
+	}
+
+	ca, err := base64.StdEncoding.DecodeString(optionalCAString)
+	if err != nil {
+		return nil, err
+	}
+	return sourcePEM(cert, key, ca)
+}
+
+// SourceDir sources certificates from directory on the host
+// Expected Certificate Name: `tls.crt`
+// Expected Private Key Name: `tls.key`
+// Expected CA Certificate Name: `ca.crt` [Optional]
+func SourceDir(dir string) (CertSource, error) {
+	if _, err := os.Stat(dir); err != nil {
+		return nil, err
+	}
+	return SourceFS(os.DirFS(dir).(fs.ReadFileFS))
+}
+
+// SourceFS sources certificates from an `io/fs.FS` abstraction
+// Expected Certificate Name: `tls.crt`
+// Expected Private Key Name: `tls.key`
+// Expected CA Certificate Name: `ca.crt` [Optional]
+//
+// If the CA Certificate is not provided, the `tls.crt` will be used as its own CA.
+func SourceFS(fsys fs.ReadFileFS) (CertSource, error) {
+	certPEM, err := fsys.ReadFile(corev1.TLSCertKey)
+	if err != nil {
+		return nil, err
+	}
+	keyPEM, err := fsys.ReadFile(corev1.TLSPrivateKeyKey)
+	if err != nil {
+		return nil, err
+	}
+	cert, err := tls.X509KeyPair(certPEM, keyPEM)
+	if err != nil {
+		return nil, err
+	}
+
+	caPEM, err := fsys.ReadFile(TLSCAKey)
+	if err != nil {
+		caPEM = certPEM
+	}
+
+	ca, err := x509.ParseCertificate(pemToDER(caPEM))
+	if err != nil {
+		return nil, err
+	}
+
+	return func(context.Context) (tls.Certificate, *x509.Certificate, error) {
+		return cert, ca, nil
```

Comment:
question: it seems like none of the certificates can be changed dynamically. Is that intended? If so, will that change anytime soon? Perhaps we overengineered this.

Reply:
Certificates from the file system can change.